stdlib/LinearAlgebra/src/blas.jl

```diff
@@ -130,6 +128,24 @@ function set_num_threads(n::Integer)
     return nothing
 end
 
+"""
+    get_num_threads()
+
+Get the number of threads the BLAS library is using.
+"""
+function get_num_threads()
+    blas = LinearAlgebra.BLAS.vendor()
+    if blas === :openblas || blas === :openblas64
+        return ccall((@blasfunc(openblas_get_num_threads), libblas), Cint, ())
+    elseif blas == :mkl
+        return ccall((:mkl_get_max_threads, libblas), Cint, ())
+    elseif Sys.isapple()
+        return Base.parse(Cint, ENV["VECLIB_MAXIMUM_THREADS"])
+    else
+        error("Unknown BLAS") # better error? return nothing
```

Comment:
I don't know whether other options are possible, but if they are, `set_num_threads` won't do anything. I think this should just return some integer, not an error. Could be `Sys.CPU_THREADS`.

Reply:
I like this part from the Zen of python
```
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
```
If we want to break it, that needs justification. I think the user should just do
```julia
try
    BLAS.get_num_threads()
...
```
BTW for `BLAS.set_num_threads` it is broken, so maybe there are good reasons, that I am not aware of.

Reply:
> This would be nice to have. I think the obvious use is
> 
> ```
> old = get_num_threads()
> @threads for i in 1:100
>     # use *
> end
> set_num_threads(old)
> ```
> 
> If it's possible to have Julia without one of the libraries handled by `set_num_threads`, then I think this should do nothing, and not give an error.

How about 
```julia
BLAS.with_num_threads(1, ignore_failure=true) do
    ...
end
```

Reply:
What's the use case where an error would be better? 

In my example above there's no ambiguity, in the sense that setting the number of threads controls speed but not correctness. And it would be pretty annoying to need two try/catch blocks. 

I presume that making `set_num_threads` give an error would also count as a breaking change (if this can every happen).

Reply:
This error helps to debug why performance is bad. Otherwise, you think BLAS threading is turned off, so you conclude the problem is with your `Threads.@threads`. Depending on the setting bad performance is a bug as severe as incorrectness.

Reply:
How about raising a warning instead of an error?

Reply:
There are lots of slow fallback methods which don't give errors. It could give a warning, I guess. 

There's some precedent for returning `nothing`, which would be fine for save-and-reset as long as there's a method like this:
```
set_num_threads(::Nothing) = set_num_threads(Sys.CPU_THREADS)
```

Reply:
I think error or `return nothing` is better than returning `Sys.CPU_THREADS`. It's better to provide a way to let the user know that `get_num_threads` failed in some sense.  I'm slightly inclined to `nothing` since it's easier to compose.  A negative side is that, _if_ we can figure out better error information, there is no way to propagate this programmatically.  But I think the chance we get to this stage is low.

Also, I don't think defining `set_num_threads(::Nothing)` is the best option.  For example, the user may have access to the number of physical CPUs (e.g., calling `lscpu`).  Also, it's trivial to write `set_num_threads(something(old, Sys.CPU_THREADS))` or `old === nothing || bset_num_threads(old)`.

Reply:
It would be a little weird to accept `nothing`, but `set_num_threads(something(old, Sys.CPU_THREADS))` also feels like easy-to-forget boilerplate for the off-chance that someone has a non-standard library (which your tests won't have). For the use I outlined, of my package re-setting it after the loop. If you know more, then it needn't get in the way.

Reply:
I don't think `set_num_threads(::Nothing) = set_num_threads(Sys.CPU_THREADS)` is weird. It'd be a good API if `Sys.CPU_THREADS` was the best choice. But we know that using the number of physical cores and respecting cgroup would be a better choice here. Having said that, I'm less opposed to defining `set_num_threads(::Nothing)` than returning `Sys.CPU_THREADS` if we clearly document that `set_num_threads(::Nothing)` will always be a best-effort and the exact number that would be set is undefined.

Reply:
There's no easy way to get the number of cores, right? 

Openblas lets me set 32 threads, while MKL restricts to 2 (on a 2/4 processor). But for both of these, you can save the default first. Not sure whether the ENV variable will always be set for Apple Accelerate. For even more exotic libraries, `set_num_threads` won't do anything anyway (right now).

Reply:
> There's no easy way to get the number of cores, right?

Yes it's not easy. I think https://github.com/JuliaParallel/Hwloc.jl does give this. But it would be nice to have this option in Base.

Reply:
I think it's straightforward in Linux. You just have to parse some text files in `/proc` and `/sys`.  Using CPUID might be the best fallback option for all OSes.  Hwloc.jl is probably not going to be in Base or stdlib.  See #33409

I think it has to be written in C (at least for now) if we want to use it also for `Threads.nthreads()`.  But nobody wants to touch it (so far).

Reply:
OK. I guess if that gets solved, this can follow later. For now I guess `Threads.nthreads()` and `div(Sys.CPU_THREADS,2)` are other possible defaults.